File: WolfDen/wwwroot/ts/Behaviours/RepairBehaviour/RepairBehaviourTemplate.ts
/// <reference path="../../Unit.ts" />

namespace TurnBasedBoardGame {
    export class RepairBehaviourTemplate implements IBehaviourTemplate {
        Class: string;

        AddBehaviour(unit: Unit): void {
            let self = this;
            let repairBehaviour = new RepairBehaviour();
            Object.assign(repairBehaviour, self);
            let unitHealth = null;
            for (let key in unit.Behaviours)
                if (IsICombatBehaviour(unit.Behaviours[key]))
                    unitHealth = unit.Behaviours[key] as IHealth;
            repairBehaviour.UnitHealth = unitHealth;
            repairBehaviour.Unit = unit;
            repairBehaviour.Game = unit.Game;
            repairBehaviour.Id = unit.Id + self.Class;
            repairBehaviour.Initialize();
            unit.Behaviours[repairBehaviour.Id] = repairBehaviour;
        }
}
}

File: WolfDen/Game/Behaviours/RepairBehaviour/RepairBehaviourTemplate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using WolfDen.Game.Interfaces;
using WolfDen.Game.Templates;

namespace WolfDen.Game.Behaviours.RepairBehaviour
{
    public class RepairBehaviourTemplate : IBehaviourTemplate
    {
        public string Class { get { return GetType().Name; } }
        public int FullRepairCost { get; set; }
        public void AddBehaviour(Unit unit)
        {
            var unitHealth = unit.Behaviours.Values.OfType<IHealth>().FirstOrDefault();
            if (unitHealth == null)
                throw new Exception("No unitHealth/combatbehaviour found while generating repair behaviour. Make sure combatbehaviour is added first.");
            var repairBehaviour = new RepairBehaviour(unit.Game, unit);
            repairBehaviour.Id = unit.Id + Class;
            repairBehaviour.Unit = unit;
            repairBehaviour.UnitHealth = unitHealth;
            repairBehaviour.FullRepairCost = FullRepairCost;
            unit.Behaviours.Add(repairBehaviour.Id, repairBehaviour);
        }
    }
}
